lib/app/upload.ex

```diff
@@ -14,53 +14,93 @@ defmodule App.Upload do
     filename: "my-awesome-image.png"
   }
   Uploads to `AWS S3` using `ExAws.S3.upload` and returns the result.
-  Note: this function follows the "Let it crash" philosophy.
-  Hence the "try rescue" block in the ApiController.
-  If the upload fails it will throw an error; we *want* that.
-  Log it and address it when it arises.
+  If the upload fails for whatever reason (invalid content type, invalid CID, request to S3 fails),
+  the an error is returned `{:error, reason}`.
   """
   def upload(image) do
     # Create `CID` from file contents so filenames are unique
-    #
-    {:ok, file_binary} = File.read(image.path)
-    file_cid = Cid.cid(file_binary)
-    file_name = "#{file_cid}.#{Enum.at(MIME.extensions(image.content_type), 0)}"
+    case File.read(image.path) do
+      {:ok, file_binary} ->
+        file_cid = Cid.cid(file_binary)
 
-    # Upload to S3
-    {:ok, body} =
-      image.path
-      |> ExAws.S3.Upload.stream_file()
-      |> ExAws.S3.upload(Application.get_env(:ex_aws, :original_bucket), file_name, acl: :public_read, content_type: image.content_type )
-      |> ExAws.request(get_ex_aws_request_config_override())
+        file_extension =
+          image.content_type
+          |> MIME.extensions()
+          |> List.first()
 
-    # Sample response:
-    # %{
-    #   body: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\n
-    #    <CompleteMultipartUploadResult xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">
-    #    <Location>https://s3.eu-west-3.amazonaws.com/imgup-original/qvWtbC7WaT.jpg</Location>
-    #    <Bucket>imgup-original</Bucket><Key>qvWtbC7WaT.jpg</Key>
-    #    <ETag>\"4ecd62951576b7e5b4a3e869e5e98a0f-1\"</ETag></CompleteMultipartUploadResult>",
-    #   headers: [
-    #     {"x-amz-id-2",
-    #      "wNTNZKt82vgnOuT1o2Tz8z3gcRzd6wXofYxQmBUkGbBGTpmv1WbwjjGiRAUtOTYIm92bh/VJHhI="},
-    #     {"x-amz-request-id", "QRENBY1MJTQWD7CZ"},
-    #     {"Date", "Tue, 13 Jun 2023 10:22:44 GMT"},
-    #     {"x-amz-expiration",
-    #      "expiry-date=\"Thu, 15 Jun 2023 00:00:00 GMT\", rule-id=\"delete-after-1-day\""},
-    #     {"x-amz-server-side-encryption", "AES256"},
-    #     {"Content-Type", "application/xml"},
-    #     {"Transfer-Encoding", "chunked"},
-    #     {"Server", "AmazonS3"}
-    #   ],
-    #   status_code: 200
-    # }
+        # Check if file `cid` and extension are valid.
+        case {file_cid, file_extension} do
```

Comment:
With these _nested_ `case` statements the complexity of this file has gone up to a `9` from a `3`. 
We don't currently have a CI check for cyclomatic complexity:
https://en.wikipedia.org/wiki/Cyclomatic_complexity

But the gist is this: if you have more than 2 nested `if`, `case` or `cond` (etc.) statements,
`then` the complexity of the function is too high and the number of branches to test becomes 6<.
This means the function needs to be split into smaller functions that are in turn easier to unit test.

Reply:
Interesting read. I've read https://www.youtube.com/watch?v=9RvUj1Psqgk&ab_channel=iProcessConsulting to get a better overview and it makes sense. I'll refactor and split validation into different private functions.

I had initially tried to use [`with`](https://www.openmymind.net/Elixirs-With-Statement/) to improve code readability but I wasn't able to properly pattern-match the happy path.

Reply:
I've split the functions and used `with` to make the code clearer (check [`beb03a1` (#86)](https://github.com/dwyl/imgup/pull/86/commits/beb03a1ecfe7f766f71b0d21ccef5e6fb8561c45), [`e307c4d` (#86)](https://github.com/dwyl/imgup/pull/86/commits/e307c4d3339f555430c41edca0fb361db5f7aa85), [`18af780` (#86)](https://github.com/dwyl/imgup/pull/86/commits/18af780c9612f145d3a42b4aff3fc21ea056f05f))